universe/Building.h

```diff
@@ -214,9 +214,8 @@ class BuildingTypeManager {
 
 private:
     BuildingTypeManager();
-    ~BuildingTypeManager();
 
-    std::map<std::string, BuildingType*> m_building_types;
+    std::map<std::string, std::unique_ptr<BuildingType>> m_building_types;
```

Comment:
Perhaps I don't understand how they work, but by thread safety, I meant that if some function's local scope includes a (valid) `shared_ptr` to an object, then that object can be assumed to exist as long as the `shared_ptr` doesn't go out of scope or get reset within that function. If the manager classes store their pointers as `shared_ptr`, and something in one thread causes them to reset all those pointers, any other copies of the `shared_ptr` in use elsewhere should keep the object from being deallocated until all copies are reset. A raw pointer wouldn't keep the object from being deallocated when the manager class resets all its pointers.

This isn't about making accessing objects thread safe, but just keeping the pointers to objects valid as long as there exists such a pointer in some scope where it could be dereferenced.